user-ops-indexer/user-ops-indexer/src/indexer/v06/indexer.rs

```diff
@@ -0,0 +1,411 @@
+            let to_block = if past_db_logs_end_block > 0 {
+                past_db_logs_end_block as u64
+            } else {
+                max(
+                    (block_number - past_rpc_logs_range) as i32 + past_db_logs_end_block,
+                    0,
+                ) as u64
+            };
+            tracing::info!(from_block, to_block, "fetching missed tx hashes in db");
+            let txs = repository::user_op::find_unprocessed_logs_tx_hashes(
+                &self.db,
+                *ENTRYPOINT_V06,
+                UserOperationEventFilter::signature(),
+                from_block,
+                to_block,
+            )
+            .await?;
+            tracing::info!(count = txs.len(), "found missed txs in db");
+            txs
+        } else {
+            Vec::new()
+        };
+
+        let recent_logs = if past_rpc_logs_range > 0 {
+            let from_block = (block_number + 1 - past_rpc_logs_range).max(0);
+            tracing::info!(
+                from_block,
+                to_block = block_number,
+                "fetching past BeforeExecution logs from rpc"
+            );
+            let filter = filter.from_block(from_block).to_block(block_number);
+            let logs = self.client.get_logs(&filter).await?;
+            tracing::info!(count = logs.len(), "fetched past BeforeExecution logs");
+            logs
+        } else {
+            Vec::new()
+        };
+
+        let stream_logs = stream::iter(recent_logs).chain(stream_logs);
+
+        // map to transactions hashes containing user ops, with deduplicated transaction hashes
+        // e.g. [A, A, B, B, B, C, C] -> [A, B, C]
+        let stream_txs = stream::iter(missed_db_txs)
+            .chain(stream_logs.filter_map(|log| async move { log.transaction_hash }))
+            .scan(H256::zero(), |prev, tx_hash| {
+                if *prev == tx_hash {
+                    future::ready(Some(None))
+                } else {
+                    *prev = tx_hash;
+                    future::ready(Some(Some(tx_hash.clone())))
+                }
+            })
+            .filter_map(|tx_hash| async move { tx_hash });
+
+        stream_txs
+            .for_each(|tx| async move {
+                if let Err(err) = &self.handle_tx(tx.clone()).await {
+                    tracing::error!(error = ?err, tx_hash = ?tx, "tx handler failed, skipping");
+                }
+                ()
+            })
+            .await;
+
+        Ok(())
+    }
+
+    async fn handle_tx(&self, tx_hash: H256) -> anyhow::Result<()> {
+        tracing::info!(tx_hash = ?tx_hash, "processing tx");
+        let tx = self
+            .client
+            .get_transaction(tx_hash)
+            .await?
+            .ok_or(anyhow!("empty transaction returned from rpc"))?;
+
+        let receipt = self
+            .client
+            .get_transaction_receipt(tx_hash)
+            .await?
+            .ok_or(anyhow!("empty receipt returned from rpc"))?;
```

Comment:
I planned to eventually use `ethers-rs` or `alloy` RPC batching for this, should be even better that way.

Reply:
Turned out the `ethers-rs` does not have proper support for batching of arbitrary requests. So will have to wait for `alloy` provider to be fully ready. Anyway, I think the benefit of JSON RPC batching in websocket connection is almost negligible, since the same tcp connection is being reused anyway.